### PR TITLE
Ts pdk

### PR DESCRIPTION
**Pull request by Phault**
Life just hit me hard, so I'm going to quickly get this PR up for you to provide feedback on or land for me if I end up going AWOL. To be clear I _do_ intend to finish this. Pretty sure I ended last weekend with it being functionally complete, and I was just planning to add docs, tests and integrating it with the CI.

I've targeted the develop-0.35 branch purely to ensure there's no hidden conflicts, but I don't expect it to be included in that release.

All the serde rename stuff in f3da954ed18155399d634146170223cf9ca7d707 was a temporary workaround to the get proper casing in the generated types. I hope you have a better solution for that part.

All the Extism declarations I hope to upstream, so you won't have to maintain it. 

There's some leftover TODO comments in there, where I'd like some input. I'll add some more context in a sec.

`binaryen` will have to also be installed in the CI pipeline, as well as extism-js unless we run `proto use` in the `package` folder.

As mentioned there's a lack of tests, and also no counterpart for the pdk-test-utils package.

For the worst offenders of missing optionality I did end up manually overriding it, to make it less horrible DX wise.

Anyway, hope you like it and sorry for dropping a 4k lines pr for you to review :p

### CODE REVIEW

File: package/src/versions/types.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import type { SemVer } from "semver";
+import type { VersionSpec } from "./resolved-spec";
+import type { UnresolvedVersionSpec } from "./unresolved-spec";
+
+// TODO: should we allow passing raw strings as version-likes?
```
Comment (Phault) [inline package/src/versions/types.ts:5]: Upside is simplicity, downside is ease of making mistakes.
It's just a matter of adding `| string` if this is something we want.

File: package/test/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  createVerifyChecksum,
+  fetchUrl,
+  getHostEnvironment,
+  getToolConfig,
+  hostLog,
+} from "@moonrepo/proto-pdk";
+
+export const register_tool = createRegisterTool(
+  (input: RegisterToolInput): RegisterToolOutput => {
+    // TODO: hostLog is a bit awkward to use
```
Comment (Phault) [inline package/test/src/index.ts:36]: I reckon something like `hostLog.trace('Hello world', optionalDataObject)` would be nicer. Alternatively could just be overloads, but that truly sucks to implement in javascript/typescript.

File: package/.prototools [hunk 3]
```diff
@@ -0,0 +1,11 @@
+node = "20.12.1"
+npm = "10.5.0"
+extism-js = "1.0.0-rc9"
+
+[plugins]
+extism-js = "source:./extism-js.toml"
+
+# TODO: add a binaryen plugin - requires a more fine grain version interpolation though
```
Comment (Phault) [inline package/.prototools:8]: Mentioned this earlier on Discord, but never got around to adding the required `{version.major}` interpolation needed.

File: package/src/versions/resolved-spec.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import { SemVer } from "semver";
+import { cleanVersionString, isAliasName } from "./utils";
+import { UnresolvedVersionSpec } from "./unresolved-spec";
+
+// TODO: Maybe make the SemVer library an internal detail and hide it as it
+// would be nice to later switch it for a tree-shakeable alternative.
```
Comment (Phault) [inline package/src/versions/resolved-spec.ts:6]: ESBuild struggles (or completely don't support?) tree-shaking CJS packages, and the semver lib is quite big. I haven't looked much into alternatives, but there are some out there.

**Comment by milesj** [comment]
Thanks for this. Will take me a bit to get through.

**Comment by Phault** [comment]
@milesj Was that an intentional close, or perhaps caused by the deletion of the 0.35 branch?

**Comment by milesj** [comment]
Looks like it was caused by merging the 0.35 branch.

**Pull request closed without merging**